File: unusual_sort.rb
#!/usr/bin/env ruby
# http://www.codewars.com/kata/un-usual-sort

def unusual_sort(list)
  digits = ["1","2","3","4","5","6","7","8","9","0"]
  uppercase, lowercase, ints, int_strings, all_ints = [],[],[],[],[]
  list.each do |e|
    if e.class == String && !digits.include?(e)
      e == e.upcase ? uppercase << e : lowercase << e
    else 
      digits.include?(e) ? int_strings << e : ints << e
    end
  end
  [uppercase,lowercase,ints,int_strings].each { |e| e.sort! }
  until ints.count == 0 || int_strings.count == 0
    ints.first <= int_strings.first.to_i ? 
      all_ints << ints.shift : all_ints << int_strings.shift
  end
  all_ints = ints.count > 0 ? ints + all_ints : all_ints + int_strings
  uppercase + lowercase + all_ints
end

sample = ["a","Z","B", 1, 5, "5", "1"]
p unusual_sort(sample)

File: unusual_sort.js
#!/usr/bin/env node
// http://www.codewars.com/kata/un-usual-sort

function unusualSort(list) {
  var uppercase = []; var lowercase = []; var ints = [];
  for (var i in list) {
    if (Number(list[i])) {
      ints.push(list[i]);
    } else {
      if (list[i].toUpperCase() === list[i]) {
        uppercase.push(list[i]);
      } else {
        lowercase.push(list[i]);
      }
    }
  }
  return uppercase.sort().concat(lowercase.sort()).concat(ints.sort())
}

var t = ["a","Z","B", 1, 5, "5", "1"];
console.log(unusualSort(t));
